### PR TITLE
Simplify registry data and add a new service data store seperated from registry #2030

### PR DESCRIPTION
**Pull request by cvictory**
## What is the purpose of the change

For issue: #2030

URL works as the carrier for Dubbo internal data transfer at runtime, and is also works as a protocol integrating with the external system: registration center, service query, service management, etc., resulting in serious coupling between the configuration and the registration center, and the data load pushed by the registration center is too large.

Keep URL unchanged at runtime, and even enrich the runtime URL; Keep the immutability of the URL during a single invoke from start to end.
Simplify Registry url as much as possible.
Refactor the way of doing service configuration, rely on Registry currently, switch to third-party Config Center while keeping compatible with old Overrides protocol.
Some data in URL are for service query purpose, we should register this part to another place instead of Registry. And it should be realized as extensible and users can decide to enable it or not.

## Brief changelog

* Define a new SPI for store service data.  ServiceStore & ServiceStoreFactory
* Remove some parameter from registry . Make it simple.  Refer to : https://github.com/apache/incubator-dubbo/issues/2030#issuecomment-418938829
* Implement the SPI with zookeeper as an example.

## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [ ] Run `mvn clean install -DskipTests` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/AbstractInterfaceConfig.java [hunk 1]
```diff
@@ -46,7 +48,11 @@
  *
  * @export
  */
-public abstract class AbstractInterfaceConfig extends AbstractMethodConfig {
+public abstract class
+
+
+
+AbstractInterfaceConfig extends AbstractMethodConfig {
```
Comment (chickenlj) [inline dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/AbstractInterfaceConfig.java]: Extra new lines here.

Comment (chickenlj) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/AbstractInterfaceConfig.java]: @cvictory

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/ReferenceBean.java [hunk 2]
```diff
@@ -149,6 +150,15 @@ public void afterPropertiesSet() throws Exception {
                 }
             }
         }
+        if (getServiceStoreConfig() == null) {
+            Map<String, ServiceStoreConfig> serviceStoreConfigMap = applicationContext == null ? null : BeanFactoryUtils.beansOfTypeIncludingAncestors(applicationContext, ServiceStoreConfig.class, false, false);
+            if (serviceStoreConfigMap != null && serviceStoreConfigMap.size() == 1) {
+                // 第一个元素
```
Comment (chickenlj) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/ReferenceBean.java]: Please use English comment.

File: dubbo-common/src/main/java/org/apache/dubbo/common/URL.java [hunk 3]
```diff
@@ -250,6 +250,47 @@ public static URL valueOf(String url) {
...
     }
 
+    public static URL valueOf(String url, String... reserveParams){
+        URL result = valueOf(url);
+        if (reserveParams == null || reserveParams.length == 0){
+            return result;
+        }
+        Map<String, String> newMap = new HashMap<String,String>(reserveParams.length);
+        Map<String, String> oldMap = result.getParameters();
+        for(String reserveParam : reserveParams){
```
Comment (diecui1202) [inline dubbo-common/src/main/java/org/apache/dubbo/common/URL.java]: code format

File: dubbo-servicedata/dubbo-servicedata-api/pom.xml [hunk 4]
```diff
@@ -0,0 +1,41 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (diecui1202) [inline dubbo-servicedata/dubbo-servicedata-api/pom.xml]: LICENSE header.

File: dubbo-servicedata/dubbo-servicedata-api/src/main/java/org/apache/dubbo/servicedata/integration/ServiceStoreService.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+import org.apache.dubbo.servicedata.support.AbstractServiceStoreFactory;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ */
```
Comment (diecui1202) [inline dubbo-servicedata/dubbo-servicedata-api/src/main/java/org/apache/dubbo/servicedata/integration/ServiceStoreService.java]: non meaningful comment

File: dubbo-servicedata/dubbo-servicedata-zookeeper/pom.xml [hunk 6]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (diecui1202) [inline dubbo-servicedata/dubbo-servicedata-zookeeper/pom.xml]: license info too.

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/ServiceBean.java [hunk 7]
```diff
@@ -219,6 +220,15 @@ public void afterPropertiesSet() throws Exception {
                 }
             }
         }
+        if (getServiceStoreConfig() == null) {
+            Map<String, ServiceStoreConfig> serviceStoreConfigMap = applicationContext == null ? null : BeanFactoryUtils.beansOfTypeIncludingAncestors(applicationContext, ServiceStoreConfig.class, false, false);
+            if (serviceStoreConfigMap != null && serviceStoreConfigMap.size() == 1) {
+                // 第一个元素
```
Comment (carryxyh) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/ServiceBean.java]: Chinese comment

**Comment by chickenlj** [comment]
* ServiceStore can replace ServiceStoreService as the entrance in RegistryConfig/ServiceConfig.
* Delete useless attributes in ServiceStoreConfig.
* Enrich method signature information, generalized inclusion and realize methods in PojoUtils may help.
* ServiceStore's HA strategy: if the server cannot be reached, ignore and try background. You can reference FailbackRegistry.
* ZK nodes written by ServiceStore should be changed to persistent
* OPS will need to retrieve the ServiceStore instance and I think you should provide the exact URL format used by ServiceStore.
* To decide which keys need to be registered to Registry, we can make it configurable by adding an item in dubbo.properties.
* Add a switch to control the URL registered to Registry: register the complete URL to the registry by default.
* Is it necessary to keep cache files locally in AbstractServiceStore? I think it's too complex, just log the error information in case of write failing can be enough.

**Pull request merged**